MariaDB/server
# [MDEV-34614] mysqlbinlog warn on EOF before GTID in --stop-position

* [x] *The Jira issue number for this PR is: [MDEV-34614](https://jira.mariadb.org/browse/MDEV-34614)*
  * Resolves https://github.com/MariaDB/server/pull/3400#discussion_r1681030120
* Follow-up of [MDEV-27037](https://jira.mariadb.org/browse/MDEV-27037)
  * Follow-up of #3400


## Description

This PR adds `mariadb-binlog` warnings for `--stop-position` GTIDs that were not reached at EOF, mainly as a follow-up to MDEV-27037 which added warnings for unreached `--stop-position` file indices and `--stop-datetime`s.

### What problem is the patch trying to solve?

`--stop-position` warnings inform possible mistakes in the input; they’re helpful for both DBAs on CLI and scripts/wrappers that check/log the output status.

MDEV-34614 enchances MDEV-27037 with per-domain GTID validation.
MDEV-27037 was considered a bug fix and so built upon the oldest supported branch (at the time) for forwardporting, but GTID range selection wasn’t a feature of its ancient base version.

<details><summary>

Unlike #3400, the code behind this feature is in the `rpl_gtid.cc` class `Binlog_gtid_state_validator` rather than directly in the `mysqlbinlog.cc` program itself via its filter inputs.
There are two ~~limitations~~ _current situations_ that lead to this decision:

</summary>

* `Binlog_gtid_state_validator` is already prompting unreached *`--start-position`s* in `--gtid-strict-mode` (errors) or `-vvv` mode (warnings).
  This may rather be a side effect. (not certain; I’m new 👶)
* Both MDEV-34614 and MDEV-27037 are considered bug fixes that will apply on older supported versions. Such tickets should have minimal changes and no refactorings.
  The GITD range filter doesn’t have an API for enumerating statuses of individual `--stop-position` GTIDs, whereas modifying `Binlog_gtid_state_validator` only requires one function to gain some optional parameters.

A “new feature” PR with unrestrained rewrites would have a more ideal Separation of Concerns, such as:
* Filters preselect which entries (row or transaction depending on the format) are processed.
* The GTID validator doesn’t care about ranges; it only validates entries (e.g., confirming GITD order) within the range given by the filters.
* The program uses the GTID filters – not the GTID validator – to warn out-of-range GTIDs to match the logic and error message of position & time ([MDEV-27037](https://jira.mariadb.org/browse/MDEV-27037)).
* For parity, the program warns out-of-range `--start-position`/`datetime`s regardless of `--gtid-strict-mode`/`-vvv`, not just `--stop-position`/`datetime`s.

</details>

### Do you think this patch might introduce side-effects in other parts of the server?

<details><summary>

To enable the `Binlog_gtid_state_validator`-based design, I had to (temporarily?) remove the *optimization* that recycles the validator object at the *first* (encounter order) `--start-position` GTID outside of `--gtid-strict-mode`/`-vvv`.

</summary>

https://github.com/MariaDB/server/blob/0e8fb977b00983d98c4c35e39bc1f36463095938/client/mysqlbinlog.cc#L1130-L1151

Note that despite this, this latter (runtime order) code queries its states again (if it’s not deleted), specifically to confirm the expectation of reaching (start) GTID states.
https://github.com/MariaDB/server/blob/0e8fb977b00983d98c4c35e39bc1f36463095938/client/mysqlbinlog.cc#L3732-L3741

Consequently, outside of `--gtid-strict-mode` and `-vvv`, the only case that enters this line is when EOF is found before __every__ *`--start-position`* GTID (rather than _any_).
(Recall that sequences of each domain have individual `--start-position`s, just that file logs are concurrent.)
One of these two designs has to be erroneous.

</details>

Optional parameters preserve the current (non-)behaviors of (non-)reporting.
Early exits from `--gtid-strict-mode` exceptions also do (should?) not prompt unreached `--start-position`s since these states may be natural for those. (We’re logging unreachability, not reachability.)


## Release Notes

*What does #3400 say for its Release Notes?… Nothing? 🤷*

Here’s a summary that covers both this and the previous PR:
> `mariadb-binlog` now warns about any `--stop-position`s and `--stop-datetime`s that didn’t apply after reading the entire log.

## How can this PR be tested?

The new test `binlog_mysqlbinlog_warn_stop_gtid` covers positive expectations (but not negative testing yet), though QA should run the entire `binlog` suite to also confirm no regressions.

## PR quality check

* *This is a new feature or a refactoring, and the PR is based against the `main` branch.*
